File: .old/Aclass_one_more_time/A1043.cpp
#include <iostream>
#include <vector>

using namespace std;

int getpost(int root, int tail, int last_root, vector<int> &pre, vector<int> &post){
    if(root > tail) return 0;
    int i = root;
    while(i < tail && pre[i + 1] < pre[root]){
        if(pre[i + 1] < last_root) return 0;
        i += 1;
    }
    if(i < tail && pre[i + 1] == pre[root]) i += 1;
    if(i < tail && pre[i + 1] < pre[root]) return 0;
    getpost(root + 1, i, -99999, pre, post);
    getpost(i + 1, tail, pre[root], pre, post);
    post.push_back(pre[root]);
}

int main() {
    int num_nodes;
    cin >> num_nodes;
    vector<int> pre(num_nodes);
    vector<int> post;
    for(int i = 0; i < num_nodes; i++){
        cin >> pre[i];
    }
    if(num_nodes == 1){
        cout << "YES" << endl;
        cout << pre[0] << endl;
        return 0;
    }
    if(pre[1] > pre[0]) {
        for(int i = 0; i < num_nodes; i++) pre[i] = -pre[i];
    }
    getpost(0, num_nodes - 1, -99999, pre, post);
    if(post.size() == num_nodes){
        cout << "YES" << endl;
        cout << abs(post[0]);
        for(int i = 1; i < num_nodes; i++) cout << " " << abs(post[i]);
    } else cout << "NO" << endl;
    return 0;
}

File: .old/Aclass_Python/A1043.py
num = int(input())
pre = list(map(int, input().split()))
if num == 1:
    print("YES")
    print(pre[0])
else:
    if pre[1] > pre[0]:
        pre = list(map(lambda x: -x, pre))
    post = []

    def getpost(root, tail, last_root):
        if root > tail:
            return
        i = root
        while i < tail and pre[i + 1] < pre[root]:
            if pre[i + 1] < last_root:
                return
            i += 1
        if i < tail and pre[i + 1] == pre[root]:
            i += 1
            if i < tail and pre[i + 1] < pre[root]:
                return
        getpost(root + 1, i, -99999)
        getpost(i + 1, tail, pre[root])
        post.append(pre[root])

    getpost(0, num - 1, -99999)
    if len(post) == num:
        print("YES")
        print(" ".join(list(map(lambda x: str(abs(x)), post))))
    else:
        print("NO")

